lib/model/CTokenListCategory.cc

```diff
@@ -380,8 +396,19 @@ CTokenListCategory::TSizeSizePr CTokenListCategory::orderedCommonTokenBounds() c
 }
 
 std::size_t CTokenListCategory::maxMatchingStringLen() const {
-    // Add a 10% margin of error
-    return (m_MaxStringLen * 11) / 10;
+    // Add a 10% margin of error if this wouldn't result in a length much
+    // longer than the base string.  The broader the category (i.e. fewer
+    // tokens that must match) the lower the tolerance of length increases.
+    // The risk is that we end up with a category with just one or two tokens
+    // that matches every message.  The max matching length is designed to
+    // prevent categories that just require a few tokens to match from
+    // matching much longer messages, but the 10% growth here can cause
+    // progressively longer messages to be included in the category over time
+    // if no cap is applied.
+    std::size_t extendedLength{std::min(
+        (m_MaxStringLen * 11) / 10,
+        m_BaseRawStringLen * std::max(m_CommonUniqueTokenIds.size() / 2, std::size_t{2}))};
+    return std::max(m_MaxStringLen, extendedLength);
```

Comment:
Alternatively, you might want to consider string length as part of the similarity metric for different categories and in this context you don't need an explicit constant, but could look at the within class variation of lengths to assess whether a candidate match is significantly outlying.

Reply:
The problem is that the definition of the category is its reverse search.  So that definition needs to be simple enough for any code elsewhere in the stack or an application using the stack to apply it.  Terms, regex and length are simple enough to be incorporated into a search or some client code.

I guess it could learn a probability distribution for the lengths of messages in the category, but that alone wouldn't solve the problem as the maximum acceptable length could still drift upwards.  There needs to be some sort of cap.  As more messages get added to the category the terms get reduced and the lengths get increased and in the limit the definition would be "everything".  We already had a constraint to stop the terms being reduced to none.  We just didn't have a constraint on the length increasing forever.

One thing we could look at in the future is writing a version of categorization that uses the data frame analytics framework.  Then the definition of a category would be "whatever messages get labelled as that category" and the way the category is determined doesn't need to worry about being able to create a simple reverse search because the reverse search is just looking for a number in the output field on the index that we labelled with the results.  And because it would see the whole data set before making any decisions there wouldn't be the problem of definitions drifting over time.

Reply:
We could potentially map some clustering procedure onto a set of hard decision criteria. For example, convex clustering which includes string length in its distance metric would produce a length range filter condition. We could discuss this a bit more offline to see what might be feasible, but I do appreciate there are extra complications.

I think this comment doesn't necessarily impose fundamental additional complications:
> There are multiple ways you could smooth out taking the decision to match over all the considerations

One high-level strategy would be to migrate to a fuzzy version of the existing logic for deciding whether a string should *ideally* create a new category, but only create that category if a suitable query can be constructed for it. This might also help assign strings to better existing categories in the case that this is not possible, assuming we rule categories out by hard constraints.

Reply:
This PR was intended to be a low risk tweak to avoid a highly undesirable situation without completely changing the way we do categorization.  I think it's best to merge it as-is, or at least quite close to what it is now.  If we want to completely change the way we do categorization in the future then that needs to be scheduled as a major project.